### PR TITLE
Support track 2 client in webapp maven plugin

### PR DESCRIPTION
**Pull request by Flanker32**
- Add track 2 and library client in webapp maven plugin
- Remove outdated codes from webapp maven plugin
- Add promption and resource group related logic to deploy mojo

### CODE REVIEW

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/DeployMojo.java [hunk 1]
```diff
@@ -78,59 +81,75 @@ private IWebApp getWebApp(final WebAppConfig config) {
...
                 .withSubscription(webAppConfig.getSubscriptionId())
                 .withAppSettings(webAppConfig.getAppSettings())
                 .commit();
+        Log.info(CREATE_WEB_APP_DONE);
+        return result;
     }
 
     private IWebApp updateWebApp(final IWebApp webApp, final WebAppConfig webAppConfig) throws AzureExecutionException {
         // update app service plan
+        Log.info(UPDATE_WEBAPP);
```
Comment (andxu) [inline azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/DeployMojo.java:127]: should provide consistent message like: Log.info(String.format(CREATE_DEPLOYMENT_SLOT, webAppConfig.getDeploymentSlotName(), webAppConfig.getAppName()));

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/AbstractWebAppMojo.java [hunk 2]
```diff
@@ -416,6 +398,30 @@ protected WebAppConfig getWebAppConfig() throws AzureExecutionException {
...
+                    convertEnvironment(azureCredentialWrapper.getEnv()));
+            final AzureResourceManager azureResourceManager = AzureResourceManager.configure()
+                    .authenticate(azureCredentialWrapper.getTokenCredential(), profile).withSubscription(getSubscriptionId());
+            return AzureAppService.auth(azureResourceManager);
+        } catch (AzureLoginException | AzureExecutionException e) {
+            throw new AzureExecutionException(e.getMessage());
+        }
+    }
+
+    protected AzureEnvironment convertEnvironment(com.microsoft.azure.AzureEnvironment environment) {
```
Comment (andxu) [inline azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/AbstractWebAppMojo.java:419]: why this conversion?

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/DeployMojo.java [hunk 3]
```diff
@@ -40,23 +41,25 @@
...
-    public static final String SLOT_SHOULD_EXIST_NOW = "Target deployment slot still does not exist. " +
-            "Please check if any error message during creation.";
+    private static final String CREATE_DEPLOYMENT_SLOT = "Creating deployment slot %s in web app %s";
+    private static final String CREATE_DEPLOYMENT_SLOT_DONE = "Successfully created the Deployment Slot.";
+    private static final String APP_SERVICE_PLAN_NOT_FOUND = "App service plan %s was not found in resource group %s";
 
     @Override
     protected void doExecute() throws AzureExecutionException {
+        // initialize library client
+        az = getOrCreateLibraryClient();
```
Comment (andxu) [inline azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/DeployMojo.java:61]: getOrCreateAzureClient

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/DeployMojo.java [hunk 4]
```diff
@@ -78,59 +81,75 @@ private IWebApp getWebApp(final WebAppConfig config) {
...
 
     private IWebApp updateWebApp(final IWebApp webApp, final WebAppConfig webAppConfig) throws AzureExecutionException {
         // update app service plan
+        Log.info(UPDATE_WEBAPP);
         final IAppServicePlan currentPlan = webApp.plan();
         final IAppServicePlan targetServicePlan = StringUtils.isEmpty(webAppConfig.getServicePlanName()) ? currentPlan :
                 az.appServicePlan(webAppConfig.getServicePlanResourceGroup(), webAppConfig.getServicePlanName());
         if (!targetServicePlan.exists()) {
-            throw new AzureExecutionException(String.format("App service plan %s was not found in resource group %s",
+            throw new AzureExecutionException(String.format(APP_SERVICE_PLAN_NOT_FOUND,
```
Comment (andxu) [inline azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/DeployMojo.java:132]: add a todo to create the service plan when it doesn't exist

Comment (Flanker32) [reply azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/DeployMojo.java]: Done

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/DeployMojo.java (hunk 3 above)
Comment (Flanker32) [reply azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/DeployMojo.java]: changed to `getOrCreateAzureAppServiceClient`

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/AbstractWebAppMojo.java (hunk 2 above)
Comment (Flanker32) [reply azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/AbstractWebAppMojo.java]: As AzureTokenWrapper still use track 1 environment

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/DeployMojo.java (hunk 1 above)
Comment (Flanker32) [reply azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/DeployMojo.java]: Fixed

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/AbstractWebAppMojo.java [hunk 5]
```diff
@@ -416,6 +398,30 @@ protected WebAppConfig getWebAppConfig() throws AzureExecutionException {
         return parser.parse();
     }
 
+    protected AzureAppService getOrCreateAzureAppServiceClient() throws AzureExecutionException {
+        try {
```
Comment (andxu) [inline azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/AbstractWebAppMojo.java:402]: protected Azure getOrCreateAzureClient() throws AzureAuthFailureException, AzureExecutionException {
        try {
            final MavenAuthConfiguration mavenAuthConfiguration = auth == null ? new MavenAuthConfiguration() : auth;
            mavenAuthConfiguration.setType(getAuthType());
            azureCredentialWrapper = MavenAuthUtils.login(session, settingsDecrypter, mavenAuthConfiguration, this.httpProxyHost, this.httpProxyPort);
            if (Objects.isNull(azureCredentialWrapper)) {
                return null;
            }

            final AzureEnvironment env = azureCredentialWrapper.getEnv();
            final String environmentName = AuthHelper.azureEnvironmentToString(env);
            if (env != AzureEnvironment.AZURE) {
                Log.prompt(String.format(USING_AZURE_ENVIRONMENT, TextUtils.cyan(environmentName)));
            }
            Log.info(azureCredentialWrapper.getCredentialDescription());
            final Proxy proxy = ProxyUtils.createHttpProxy(getHttpProxyHost(), Integer.toString(getHttpProxyPort()));
            final Azure tempAzure = Azure.configure()
                    .withProxy(proxy)
                    .authenticate(azureCredentialWrapper.getAzureTokenCredentials()).withDefaultSubscription();
            final PagedList<Subscription> subscriptions = tempAzure.subscriptions().list();
            subscriptions.loadAll();
            final String targetSubscriptionId = getTargetSubscriptionId(tempAzure, subscriptions);
            checkSubscription(subscriptions, targetSubscriptionId);
            azureCredentialWrapper.withDefaultSubscriptionId(targetSubscriptionId);
            return AzureClientFactory.getAzureClient(azureCredentialWrapper, getUserAgent(), getHttpProxyHost(), getHttpProxyPort());
        } catch (AzureLoginException | IOException e) {
            throw new AzureAuthFailureException(e.getMessage());
        }
    }

File: azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/AbstractWebAppMojo.java [hunk 6]
```diff
@@ -419,10 +418,105 @@ protected WebAppConfig getWebAppConfig() throws AzureExecutionException {
         final SchemaVersion version = SchemaVersion.fromString(getSchemaVersion());
         final AbstractConfigurationValidator validator = version == SchemaVersion.V2 ?
                 new V2ConfigurationValidator(this) : new V1ConfigurationValidator(this);
-        final AbstractConfigParser parser = version == SchemaVersion.V2 ? new V2ConfigParser(this, validator) : new V2ConfigParser(this, validator);
+        final AbstractConfigParser parser = version == SchemaVersion.V2 ? new V2ConfigParser(this, validator) : new V1ConfigParser(this, validator);
         return parser.parse();
     }
 
+    protected AzureAppService getOrCreateAzureAppServiceClient() throws AzureExecutionException {
```
Comment (andxu) [inline azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/AbstractWebAppMojo.java:425]: shall we delete getOrCreateAzureClient in class AbstractAzureMojo in future?

Comment (andxu) [reply azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/AbstractWebAppMojo.java:425]: selectSubscription checkSubscription are very similar with the same class in  AbstractAzureMojo, or shall we move getOrCreateAzureAppServiceClient to AbstractAppServiceMojo and redefine selectSubscription/checkSubscription in  AbstractAzureMojo

Comment (Flanker32) [reply azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/AbstractWebAppMojo.java:425]: Yes, there is an issue that our codes for get azure client and select subscription uses models from SDK, and most maven plugin still uses the track one version, while webapp maven plugin uses the model from track2. This is the workaround that create another implement with track2 sdk for webapp only, and we will finally migrate all maven tooling to track2 and replace the duplicate implements.

**Review by andxu — APPROVED** [review]

**Pull request merged**